File: examples/coagents-routing/agent-python/agent/my_agent/joke_agent.py
"""Test Joker Agent"""

from typing import Any, cast

from langgraph.graph import StateGraph, END
from langgraph.graph import MessagesState
from langgraph.checkpoint.memory import MemorySaver
from langchain_core.runnables import RunnableConfig
from langchain_core.messages import SystemMessage, ToolMessage


from copilotkit.langchain import copilotkit_customize_config, copilotkit_exit
from pydantic import BaseModel, Field
from my_agent.model import get_model

class JokeAgentState(MessagesState):
    """Joke Agent State"""
    joke: str
    model: str

class make_joke(BaseModel): # pylint: disable=invalid-name
    """
    Make a funny joke.
    """
    the_joke: str = Field(..., description="The joke")


async def joke_node(state: JokeAgentState, config: RunnableConfig):
    """
    Make a joke.
    """

    config = copilotkit_customize_config(
        config,
        emit_messages=True,
        emit_intermediate_state=[
            {
                "state_key": "joke",
                "tool": "make_joke",
                "tool_argument": "the_joke"
            },
        ]
    )

    system_message = "You make funny jokes."

    joke_model = get_model(state).bind_tools(
        [make_joke],
        tool_choice="make_joke"
    )

    response = await joke_model.ainvoke([
        SystemMessage(
            content=system_message
        ),
        *state["messages"]
    ], config)

    tool_calls = getattr(response, "tool_calls")

    joke = tool_calls[0]["args"]["the_joke"]

    await copilotkit_exit(config)

    return {
        "messages": [
            response,
            ToolMessage(
                name=tool_calls[0]["name"],
                content=joke,
                tool_call_id=tool_calls[0]["id"]
            )
        ],
        "joke": joke,
    }

workflow = StateGraph(JokeAgentState)
workflow.add_node("joke_node", cast(Any, joke_node))
workflow.set_entry_point("joke_node")

workflow.add_edge("joke_node", END)
memory = MemorySaver()
joke_graph = workflow.compile(checkpointer=memory)

File: examples/coagents-routing/agent-js/src/joke-agent.ts
/**
 * Test Joker Agent
 */

import { tool } from "@langchain/core/tools";
import { z } from "zod";
import { RunnableConfig } from "@langchain/core/runnables";
import {
  copilotKitCustomizeConfig,
  copilotKitExit,
} from "@copilotkit/sdk-js/langchain";
import { SystemMessage, ToolMessage } from "@langchain/core/messages";
import { getModel } from "./model";
import { END, MemorySaver, StateGraph, Annotation } from "@langchain/langgraph";

import { CopilotKitStateAnnotation } from "@copilotkit/sdk-js/langchain";

// Define the JokeAgentState annotation, extending MessagesState
export const JokeAgentStateAnnotation = Annotation.Root({
  model: Annotation<string>(),
  joke: Annotation<string>(),
  ...CopilotKitStateAnnotation.spec,
});

export type JokeAgentState = typeof JokeAgentStateAnnotation.State;

const make_joke = tool(() => {}, {
  name: "make_joke",
  description: "Make a funny joke.",
  schema: z.object({
    the_joke: z.string().describe("The joke"),
  }),
});

export async function joke_node(state: JokeAgentState, config: RunnableConfig) {
  /**
   * Make a joke.
   */

  config = copilotKitCustomizeConfig(config, {
    emitMessages: true,
    emitIntermediateState: [
      {
        stateKey: "joke",
        tool: "make_joke",
        toolArgument: "the_joke",
      },
    ],
  });

  const system_message = "You make funny jokes.";

  const joke_model = getModel(state).bindTools!([make_joke], {
    tool_choice: "make_joke",
  });

  const response = await joke_model.invoke(
    [new SystemMessage({ content: system_message }), ...state.messages],
    config
  );

  const tool_calls = response.tool_calls;

  const joke = tool_calls?.[0]?.args.the_joke;

  await copilotKitExit(config);

  return {
    messages: [
      response,
      new ToolMessage({
        name: tool_calls?.[0]?.name,
        content: joke,
        tool_call_id: tool_calls?.[0]?.id!,
      }),
    ],
    joke: joke,
  };
}

const workflow = new StateGraph(JokeAgentStateAnnotation)
  .addNode("joke_node", joke_node)
  .setEntryPoint("joke_node")
  .addEdge("joke_node", END);

const memory = new MemorySaver();

export const jokeGraph = workflow.compile({
  checkpointer: memory,
});
